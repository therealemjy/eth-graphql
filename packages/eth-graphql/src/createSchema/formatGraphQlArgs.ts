import { ArgumentNode } from 'graphql';

const formatGraphQlArgs = (args: ReadonlyArray<ArgumentNode>) =>
  args.reduce<ReadonlyArray<string | boolean>>(
    (accArguments, argument) =>
      // TODO: handle objects and arrays
      'value' in argument.value ? [...accArguments, argument.value.value] : accArguments,
    [],
  );

export default formatGraphQlArgs;
