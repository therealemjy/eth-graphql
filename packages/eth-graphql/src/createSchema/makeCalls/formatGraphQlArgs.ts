import { JsonFragmentType } from '@ethersproject/abi';

import formatToEntityName from '../../utilities/formatToEntityName';
import { ContractCallArgs } from './types';

const formatGraphQlArgs = ({
  callArguments,
  abiItemInputs,
}: {
  callArguments: ContractCallArgs;
  abiItemInputs: readonly JsonFragmentType[];
}) =>
  // Go through ABI item to format call args and return them in the correct
  // order
  abiItemInputs.map((component, componentIndex) => {
    const argumentName = formatToEntityName({
      name: component.name,
      index: componentIndex,
      type: 'arg',
    });

    const argumentValue = callArguments[argumentName];
    return argumentValue;
  });

export default formatGraphQlArgs;
