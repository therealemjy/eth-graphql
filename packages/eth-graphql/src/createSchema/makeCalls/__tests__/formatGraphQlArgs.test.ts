import { ArgumentNode, Kind, ValueNode } from 'graphql';

import formatGraphQlArgs from '../formatGraphQlArgs';

describe('createSchema/makeCalls/formatGraphQlArgs', () => {
  it.each([Kind.NULL, Kind.VARIABLE])(
    'throws an error if one of the argument values is of the kind %s',
    kind => {
      const args: ArgumentNode[] = [
        {
          kind: Kind.ARGUMENT,
          name: {
            kind: Kind.NAME,
            value: 'fake argument name',
          },
          value: {
            kind,
            value: 'fake value',
          } as ValueNode,
        },
      ];

      try {
        formatGraphQlArgs(args);

        throw new Error('formatGraphQlArgs should have thrown an error but did not');
      } catch (error) {
        expect(error).toMatchSnapshot();
      }
    },
  );

  it('converts argument nodes to valid arguments', () => {
    const args: ArgumentNode[] = [
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 0',
        },
        value: {
          kind: Kind.LIST,
          values: [
            {
              kind: Kind.LIST,
              values: [
                {
                  kind: Kind.STRING,
                  value: 'fake string 0',
                },
              ],
            },
          ],
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 1',
        },
        value: {
          kind: Kind.OBJECT,
          fields: [
            {
              kind: Kind.OBJECT_FIELD,
              name: {
                kind: Kind.NAME,
                value: 'fake name 2',
              },
              value: {
                kind: Kind.OBJECT,
                fields: [
                  {
                    kind: Kind.OBJECT_FIELD,
                    name: {
                      kind: Kind.NAME,
                      value: 'fake name 3',
                    },
                    value: {
                      kind: Kind.STRING,
                      value: 'fake string 1',
                    },
                  },
                ],
              },
            },
          ],
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 4',
        },
        value: {
          kind: Kind.STRING,
          value: 'fake string 2',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 5',
        },
        value: {
          kind: Kind.BOOLEAN,
          value: true,
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 6',
        },
        value: {
          kind: Kind.INT,
          value: '100',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 7',
        },
        value: {
          kind: Kind.FLOAT,
          value: '0.1',
        },
      },
      {
        kind: Kind.ARGUMENT,
        name: {
          kind: Kind.NAME,
          value: 'fake name 7',
        },
        value: {
          kind: Kind.ENUM,
          value: '0',
        },
      },
    ];

    const res = formatGraphQlArgs(args);

    expect(res).toMatchSnapshot();
  });
});
