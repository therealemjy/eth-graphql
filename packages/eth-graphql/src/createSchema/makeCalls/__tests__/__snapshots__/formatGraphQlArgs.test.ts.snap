// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSchema/makeCalls/formatGraphQlArgs converts argument nodes to valid arguments 1`] = `
[
  [
    [
      "fake string 0",
    ],
  ],
  [
    [
      "fake string 1",
    ],
  ],
  "fake string 2",
  true,
  "100",
  "0.1",
  "0",
]
`;
